backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/DatabaseDocumentNumberRecyclingService.java

```diff
@@ -75,45 +77,110 @@ public Optional<String> addForRecycling(
               .abbreviation(documentationOfficeAbbreviation)
               .build();
 
-      return Optional.of(repository.save(deleted).getDocumentNumber());
+      repository.save(deleted);
 
     } catch (Exception e) {
       log.warn("Won´t reuse the document number", e);
-      return Optional.empty();
+      throw e;
     }
   }
 
   /**
-   * Returns a reusable document number
+   * Returns a reusable document number from the saved entries. It will remove it from the list in
+   * case it does not meet the function condition.
    *
-   * @param documentationOfficeAbbreviation
-   * @param year
-   * @return
+   * @param documentationOfficeAbbreviation the court abbreviation as it is saved in the deleted
+   *     document numbers table.
+   * @param year the year to search for, default is the current year.
+   * @return a document number to reuse.
    */
-  public Optional<String> findDeletedDocumentNumber(
-      String documentationOfficeAbbreviation, Year year) {
+  @Override
+  public String recycleFromDeletedDocumentationUnit(
+      String documentationOfficeAbbreviation, Year year)
+      throws DocumentNumberPatternException, DocumentNumberRecyclingException {
 
     var optionalDeletedDocumentationUnitDTO =
         repository.findFirstByAbbreviationAndYear(documentationOfficeAbbreviation, year);
 
     try {
       if (optionalDeletedDocumentationUnitDTO.isEmpty()) {
-        return Optional.empty();
+        throw new DocumentNumberRecyclingException("No documentation number to reuse");
       }
 
       var deletedDocumentationUnitDTO = optionalDeletedDocumentationUnitDTO.get();
 
-      if (!documentNumberPatternConfig.hasValidPattern(
-          documentationOfficeAbbreviation, deletedDocumentationUnitDTO.getDocumentNumber())) {
-        throw new DocumentNumberPatternException("Pattern is invalid");
+      if (documentationUnitRepository
+          .findByDocumentNumber(deletedDocumentationUnitDTO.getDocumentNumber())
+          .isPresent()) {
+        throw new DocumentNumberRecyclingException(
+            "Saved documentation number is currently in use");
```

Comment:
Hey @Nadav-B - This is has not directly been your task, but I just saw that in this case, we could(should) probably remove the entry from DatabaseDeletedDocumentationIdsRepository

Reply:
@HPrinz In line [125](https://github.com/digitalservicebund/ris-backend-service/blob/65a7432edfa044f3368323dfa727a1db08bd087b/backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/DatabaseDocumentNumberRecyclingService.java#L126-L126), it will end up removing the deleted document in the catch exception.
```
 optionalDeletedDocumentationUnitDTO.ifPresent(
          deletedDocumentationUnitDTO ->
              repository.deleteById(deletedDocumentationUnitDTO.getDocumentNumber()));
```

Reply:
Oh, haven't seen that, thank you! 🙏